src/charm.py

```diff
@@ -13,92 +13,186 @@
-        this method.
-
-        Learn more about config at https://juju.is/docs/sdk/config
+        if not self._snap_path_set:
+            try:
+                self._snap_path = str(self.model.resources.fetch("exporter-snap"))
+                # Don't return path to empty resource file
+                if not os.path.getsize(self._snap_path) > 0:
+                    self._snap_path = None
+            except ModelError:
+                self._snap_path = None
+            finally:
+                self._snap_path_set = True
+
+        return self._snap_path
+
+    def get_controller_ca(self) -> str:
+        """Get CA certificate used by targeted Juju controller.
+
+        CA certificate can be directly configured by `controller-ca` option, if it is, the value is
+        directly returned by this method. If it is not defined, a CA cert used by the controller
+        that deploys this unit will be returned.
         """
-        # Fetch the new config value
-        log_level = self.model.config["log-level"].lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
+        explicit_cert = self.config.get("controller-ca", "")
+        if explicit_cert:
```

Comment:
It seems we use either explicit_cert or agent_conf_path, mutually exclusive. 
I would like to code it more clear. just a suggest, not mandatory

if case_a :
    .....
    ca_cert = get_data_from_explicit_cert()
else:
    ca_cert = get_data_from_agent_conf_path()

 return ca_cert

Reply:
I think I recently had the opposite situation where I chose the "single return" approach and reviewers argued that the "early return" approach is preferable. Eventually too many people voted in favor of early return that I changed it :D 

([found the review conversation](https://github.com/canonical/juju-verify/pull/73#pullrequestreview-1147212842))

Reply:
btw, I'm also generally member of team "Single Return Statement Is Better". I'd just hate to change this and have other reviewers demanding the change back. So I think we need some consensus before committing.

(@rgildein @esunar thoughts?)

Reply:
I agree with Martin, we should stick with one approach. For me is more readable to use multiple returns and not assigned output to variable. The default output is at the bottom and any if cases on top of the function. e.g.

```python
if case_a:
    ...
    return get_data_from_explicit_cert()
elif case_b:
    ...
    return ...

...
return get_data_from_agent_conf_path()
```

Reply:
I would prefer separate functions:
`cert = get_explicit_cert() if explicit_cert else get_cert_from_agent_conf()
if not cert:
    raise RuntimeError()  

return cert`

Reply:
I will not fight for it either :)